### PR TITLE
reverse string exercise

### PR DESCRIPTION
**Pull request by eliaahadi**
I added new reverse string exercise, I'm happy to edit it.

### CODE REVIEW

**Comment by matthewmorgan** [comment]
@eliaahadi my earlier comments on this PR seem to have mysteriously disappeared.

I did see a comment from you last night asking how to move forward on this PR, and I think I had not explained myself clearly.  At the risk of repeating myself and boring you, here is what I was trying to say:

1. We need to get permission from the author of the article on Medium to reproduce her work here.  I know there have been cases in the past where authors did not want their work used.  If this is in fact all your original work, simply based on an idea she presented, maybe it's OK but it would still be polite to get her blessing.  @kytrinyx would you care to comment further on this?  I don't know what other considerations you might have.

2. We don't normally add new exercises to an individual language track unless they're first added to the canonical problem specifications repo, here: https://github.com/exercism/problem-specifications

The process of adding a new exercise normally involves specifying:
- a README 
- canonical test data  
- metadata
You've provided the first in this PR, so that work could easily be moved over to a PR on the problem-specifications repository.

If we can get permission on the first point, then it would make sense to open a PR in the problem-specifications repo to add the exercise _generally_ to exercism.  _Then_ we can accept a PR to add it specifically to this language.

Does that make sense?  Please let me know if I'm being unclear in any way.

**Comment by eliaahadi** [comment]
@matthewmorgan this is a common interview question, alittle more difficult than hello world and has many solutions on the internet. I just googled a random solution but if you feel like we need to confirm it with that specific author, I can try reaching out to her for her permission or I can just find another article and author that may be easier to reach.

I also have those files you mentioned, I tried to submit it to this PR but was advised that I don't need them at the moment.

**Comment by matthewmorgan** [comment]
@eliaahadi I understand that it's a common question, I agree with that.  I feel like I'm doing an ineffective job of communicating why we can't move forward on this PR yet, so let me try again:

A. You cite the specific article from an author on Medium in the README.  Does your PR contain original work from someone other than yourself that you haven't cited? If it does, you need to remove that, or get permission from the author.  If, on the other hand, all you are doing is mentioning that she wrote an article that gave you an idea, and that you're not actually using her work, then everything is OK and it's fine to mention the article.

B. Your contribution would be a **completely new exercises that have never been implemented on exercism.io**.  Because of that, it will **first** need to be added to the problem-specifications repo.   When that PR is approved and merged, we can then accept **this** PR for an _implementation_ of the problem in _this_ track.

Please take the following steps:

1. Submit a PR with the three files I mentioned above in [this repo](https://github.com/exercism/problem-specifications).
2.  When the PR in problem-specifications is approved, we can continue with this one.

We are happy to have your contribution, but this is our process.

Is that clear, or is there anything else I can explain?

**Comment by eliaahadi** [comment]
Ok thanks for the information. 

For #1, this PR doesn't contain any original work, the article I referred to just shows various ways to solve the problem. I'd think it's just fine to mention the article.

For #2, I did now create a new PR referred [here at #960](https://github.com/exercism/problem-specifications/pull/960). It looks like that PR checks have already failed with error "The Travis CI build failed". I'm not quite sure how to debug.

Thoughts?

**Comment by matthewmorgan** [comment]
OK @eliaahadi I'm glad we're on the same page now!

I am not a maintainer on the problem-specification repo, but I am happy to take a look and see if I can understand why the build is failing.

**Comment by matthewmorgan** [comment]
@eliaahadi When I look at the [build log](https://travis-ci.org/exercism/problem-specifications/builds/289224162#L483) at line 483, I see an error that is helpful.

Apparently for this repo you need to call your `README.md` file by a different name: `description.md`.

Also, I noticed that your description file contains information about how to run tests etc that is specific to the JavaScript track.  You don't want that in the _generic_ `description.md` I don't think.  

I would suggest checking the [contributor guide](https://github.com/exercism/problem-specifications/blob/master/CONTRIBUTING.md) for instructions, and if you don't find the information clear, you should ask on that PR, so one of the maintainers can help.  That's what I would do if I were making a PR there!

**Comment by kytrinyx** [comment]
It sounds like you've gotten things mostly straightened out.

We have some documentation about adding brand new exercises: https://github.com/exercism/docs/blob/master/you-can-help/make-up-new-exercises.md
I realize it may be a bit late for that, but it may have some context that helps.

In terms of including content from other locations, there's some discussion about this here: https://github.com/exercism/discussions/issues/188

The TL;DR is that if there is an exercise that is commonly used, then we can implement our own description and implementations of it, and refer to the source as "common interview question" (or whatever). We can also cite a source of inspiration (e.g. for the 99 bottles of beer song problem we refer to Chris Pine's book, but could just as easily have referenced http://www.99-bottles-of-beer.net/).

We _cannot_ use other people's copyrighted material. If other people have written code that we include, it has to be released under a compatible license. I prefer to avoid that entirely, because I am not a lawyer, and it's very complicated.

For some problems we make up a similar problem. E.g. instead of "fizzbuzz" we have "raindrops". They're not identical but they have similarities in structure.

**Comment by matthewmorgan** [comment]
@eliaahadi I'm glad to see your contribution was merged in to the problem-specifications repo.  Congratulations!

Do you intend to implement this exercise here?  Please let me know, otherwise we'll close this PR for now.

**Comment by eliaahadi** [comment]
I tried to push the 3 files required here, but I got permission denied error below. Thoughts? Thanks!

```
Elias-MacBook-Pro:reverse-string eliaahadi$ git push
remote: Permission to exercism/javascript.git denied to eliaahadi.
fatal: unable to access 'https://github.com/exercism/javascript.git/': The requested URL returned error: 403
```

**Comment by matthewmorgan** [comment]
Hi @eliaahadi glad you're moving ahead with this PR!

I see that you have some test failures and some linting issues.  Would you please resolve those and I would be happy to review the code then.

Let me know if you need assistance resolving any of the issues.

**Comment by eliaahadi** [comment]
@matthewmorgan I'm not quite sure how to check for the lines and edit exactly where they are from the latest commit. Thoughts?

**Comment by matthewmorgan** [comment]
@eliaahadi I would start by having a look at the [Travis log](https://travis-ci.org/exercism/javascript/builds/308151499?utm_source=github_status&utm_medium=notification)

Near the bottom there is a section that show your tests failing, and explain why they are failing-- IE, `match` is not a function.  Your `ReverseString` object _has no_ method called `match`.

Does that help?

**Comment by eliaahadi** [comment]
@matthewmorgan the latest failure says " Expected 'Ramen' to equal 'nemaR'.". I'm not sure how to test and fix this, thoughts? Thanks.

**Comment by matthewmorgan** [comment]
What do you think that test result is telling you?

**Comment by eliaahadi** [comment]
``` xit('a capitalized word', function () {
    var subject = new ReverseString('Ramen');
    var matches = subject.matches('nemaR');

    expect(matches).toEqual('nemaR');
  });```
Subject using ReverseString function ('Ramen') as argument. The variable matches using the function subject which is from ReverseString. I'd expect subject function which reversing the string Ramen to match 'nemaR' from the "expect" line. Thoughts?

I referred to this to write my tests.
[anagram](https://github.com/exercism/javascript/blob/master/exercises/anagram/anagram.spec.js)

**Comment by matthewmorgan** [comment]
1. What does the `matches` function do to its input?
2. What is the input?
3.  Therefore, what is the output of `matches`?

**Comment by eliaahadi** [comment]
1. What does the matches function do to its input? I expect sbuject.matches to reverse 'Ramen' and match 'nemaR'.
2. What is the input? Input is 'Ramen'
3. Therefore, what is the output of matches? Output after going through ReverseString function, gives, 'nemaR'. 

I'm not so familiar with writing tests which is why I referred to the anagram example.

**Comment by matthewmorgan** [comment]
2. The input to `matches` is not `Ramen`.  It is `nemaR`.

**Comment by eliaahadi** [comment]
Ok I think those errors were fixed in latest commit. Now I have spacing issues. Is there a way to see exactly where they are? Maybe an online ESLINT to check?

**Comment by matthewmorgan** [comment]
Good progress.

The log tells you where the errors are.  For example:

```bash
/home/travis/build/exercism/javascript/exercises/reverse-string/example.js
   4:1   error    Expected indentation of 2 spaces but found 1 tab  indent
```

This indicates that the file that contains the error is `example.js`.  The numbers in the next line `4:1` mean line 4, character 1 is where the problem is, and so on...

**Comment by eliaahadi** [comment]
Ok I fixed spacing issues, but now there's other errors below. Thoughts? Do I need to make another PR for the config file?

image

**Comment by matthewmorgan** [comment]
@eliaahadi you don't need to make another PR.  What do you think line 1338 means?

**Comment by eliaahadi** [comment]
It sounds like I need to edit config.json, no? PR #418 was closed and had config.json, do we need to do something with config.json?

**Comment by matthewmorgan** [comment]
Yes, you need to add the exercise to `config.json`.  That's correct!

Then, we should make a small change to the API for the exercise, just to simplify things and make the implementation easier for the students to understand.  But we can discuss that after the `config.json` changes are complete.

**Comment by eliaahadi** [comment]
Ok I made new PR for the config file at #464 . Please let me know what's next.

**Comment by matthewmorgan** [comment]
@eliaahadi Please do not open a new PR.  Please add the `config.json` changes to _this_ PR.

**Comment by eliaahadi** [comment]
Ok last check finally passed, anything else to do on this to get it added to the production website?

**Comment by matthewmorgan** [comment]
Great!

Now we need to make a change to the API to clean things up a bit.

The current state of the API:
1. We have a class `ReverseString` that takes a string input, and doesn't do anything with it.
2. We have a method on that class called `matches` which takes a second string, reverses it, and returns a string.

_Neither_ of these things do what one would expect them to do given their names, which is confusing.  Instead, what if:

A. `ReverseString` takes in a string and reverses it, storing the reversed string internally.
B. `matches` take a string to _compare_ to the reversed string, and returns `true` if the two strings match, or `false` if they do not.

 Does that make sense to you?

**Comment by eliaahadi** [comment]
Hmm ok, the ReverseSTring function is this, correct?
```
function ReverseString(string) {
  var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
  return revString;
};
```
I'm not sure what to do with the matches function though. Thoughts?

**Comment by matthewmorgan** [comment]
Well,
1. `ReverseString` is the class constructor.  It does not need to return the value, just store it internally.
2.  `matches()`: please see my earlier comments.  What inputs should matches take, and what should it return?

**Comment by eliaahadi** [comment]
In example.js, I tested this in console, 

```
'use strict';
function ReverseString(string) {
   var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
}



ReverseString.prototype.matches = function (string, revString) {

   var matchesRevString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    matchesRevString += string[i];
  }
  if (matchesRevString === revString){
  	return true;
  }
  return false;
};

module.exports = ReverseString;
```
Does that work?

**Comment by matthewmorgan** [comment]
@eliaahadi have you tried testing it against the test files?

**Comment by eliaahadi** [comment]
I would have to commit it in order to go through the Travis CI. Is that what you mean? Or is there another way?

**Comment by matthewmorgan** [comment]
The preferred way is to run the tests locally.  Do you know how to do that, or do you need help with that?

**Comment by eliaahadi** [comment]
Sure, does that mean I should download Travis CI software somewhere and try to do those tests? Or something else?

**Comment by matthewmorgan** [comment]
No, nothing so complicated.  You can simply run the tests locally for a single exercise, as described in the [README.md](https://github.com/exercism/javascript/blob/master/README.md) for this track.

**Comment by eliaahadi** [comment]
Ok I created a "reverse-string.js" file to test since that's what the hello-world had. The file contents are:

```
var ReverseString = function (string) {};

ReverseString.prototype.reverse = function(string) {
   var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
  console.log(revString);
}

module.exports = ReverseString;
```
In the folder, I get subject.matches isn't a function again. The xit tests are disabled afterwards. I'm not sure how to proceed next. Thoughts?

image

**Comment by matthewmorgan** [comment]
On you ReverseString class, _is there_ a function called `matches()`?  I don't see one in `reverse-string.js`.

**Comment by eliaahadi** [comment]
Ok I rewrote the function this way for reverse-string.js. It seemed to pass the tests. 

```
'use strict';


function ReverseString(string) {
  this.string = string;
}

ReverseString.prototype.matches = function (string) {
  var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
  return revString;
};

module.exports = ReverseString;
```

**Comment by matthewmorgan** [comment]
I'm glad to hear it passes the tests, but we're still not ready to call this PR finished.   The basic problem is simple: you are doing the reversing of the string in your `matches` method.

**A function called `ReverseString` should do one thing: reverse a string **

You really do not need any other functions to make this work.  Just fix that one.

**Comment by eliaahadi** [comment]
If I just fix the function ReverseString, it passes all tests. However, you previously mentioned to update the matches function to test if input string and revString are equal, return true, otherwise return false. If I change to boolean output, the tests fail since the reverse-string.spec.js matches test against actual values, not booleans. Thoughts?

```
'use strict';

function ReverseString(string) {
    var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
}

ReverseString.prototype.matches = function (string) {
  var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
  console.log(revString);
  return revString;
};

module.exports = ReverseString;
```

**Comment by matthewmorgan** [comment]
@eliaahadi I think I should be more explicit about what needs to change about the PR for it to be ready to merge.  My plan had been to help you refactor your solution, first by fixing your two functions, then by simplifying the solution, but I can see that I confused you, so sorry about that.

What I was suggesting earlier is that you have two functions, each of which:
1. is clearly named
2. does one thing
3. the name of the function and the action of the function are similar.

Let's try a different approach.  This exercise can, and really should, be simplified to _one_ function: `ReverseString`.  This function should take a string, and return the reverse of that string.  You don't need a `matches` function at all.  

Right now, you have a `reverseString` that doesn't return anything, and a `ReverseString.matches` function that doesn't do any matching, but rather returns a reversed string.  

Please write a `ReverseString` (capital R) class that 
1. Takes a string
2. Returns the reversed string

Remove all of the other code in your example solution please.

After that we will fix your tests.

**Comment by eliaahadi** [comment]
Ok I apologize again for not understanding this better. So just this function for the reverse-string.js?

```
'use strict';

function ReverseString(string) {
  var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
  return revString;
}

module.exports = ReverseString;
```

**Comment by matthewmorgan** [comment]
@eliaahadi please, no need to apologize!  We're all here to learn-- that's the point!

That function does what it says, which is good. Now the tests will need to be updated to reflect this change in the API of the business class.

Two things:
1. Since your new function is not really a class constructor, but a simple function, you no longer need to call `new`.  Just pass the input to the function and capture the result in a variable for testing.  I suggest a variable called `actual`.
2. You don't need the `matches` function.  You can just compare the _expected_ reversed string with whatever _actual_ value the `ReverseString` gave you using `toEqual()`.

Let me know if you need further assistance.

**Comment by eliaahadi** [comment]
Ok, I'm not sure this is what you mean. So in reverse-string.spec.js for testing, it should be this below?

```
describe('Reverse a string', function (input) {
  var reversedString = input;

  it('reversing a string', function () {
    expect(ReverseString(string)).toEqual(actual);
  });
});
```

**Comment by matthewmorgan** [comment]
Your example above compares the output of `ReverseString` to the _input_.  

Do you expect the output of `ReverseString` to equal its input?

**Comment by eliaahadi** [comment]
So input doesn't seem to do anything, it just is passed as argument. I'd just think to do the same reverse string test inside and then compare. Does this make sense below?

```
var ReverseString = require('./reverse-string');

///Just pass the input to the function and capture the result in a variable for testing. I suggest a variable called actual.
describe('Reverse a string', function (input) {
  var actual = '';
 for (var i = input.length - 1; i >= 0; i--) {
    actual += input[i];
  }
  it('reversing a string', function () {
    //compare the expected reversed string with actual value the ReverseString gave you using toEqual().
    expect(ReverseString(string)).toEqual(actual);
  });
});
```

Thoughts?

**Comment by matthewmorgan** [comment]
@eliaahadi I did not mean to imply that you should pass an input to the _test_ method (`it()`).  I can't think of a situation in which you would want to do that.  

What I meant was: you are passing an _input_ to `ReverseString`.  IE, What do you expect the _output_ of `ReverseString` to be?  Compare the expected output to the actual output.

Each tests should follow a simple pattern that is easy to understand.  For example:

```javascript
it('can test a backwards cat', function() {
  var expected = 'tac';                         // the correct result is visible right in the test
  var actual = ReverseString('cat');    // call your business function

  expect(actual).toEqual(expected);  // compare the two values
});
```

**Comment by eliaahadi** [comment]
Ok that makes more sense. The output of reverse string should be the reverse of a string. For example, in reverse-string

```
'use strict';

function ReverseString(string) {
  var revString = '';
  for (var i = string.length - 1; i >= 0; i--) {
    revString += string[i];
  }
  return revString;
}

module.exports = ReverseString;
```
```
console.log(ReverseString("cat")); // -> tac
```

Here's the screenshot of output from the console of Chrome.
image

So in the reverse-string.spec.js file, the code you written is as follows:

```
describe('ReverseString', function () {
it('can reverse a cat', function() {
  var expected = 'tac';                // the correct result is visible right in the test
  var actual = ReverseString('cat');   // call your business function

  expect(actual).toEqual(expected);    // compare the two values
});
```

Is that right?

**Comment by eliaahadi** [comment]
I updated the reverse-string.spec.js with the new example.js that passes all the tests. Can you please review and let me know your thoughts @matthewmorgan? Thank you and happy holidays!

File: exercises/reverse-string/reverse-string.spec.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+var reverseString = require('./reverse-string');
+
+describe('ReverseString', function () {
+  it('empty string', function () {
+    var expected = '';
+    var actual = reverseString('');
+    expect(actual).toEqual(expected);  
+  });
```
Comment (matthewmorgan) [inline exercises/reverse-string/reverse-string.spec.js]: Please add a blank line between this `it()` block and the next.

File: exercises/reverse-string/reverse-string.spec.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  it('a word', function () {
+    var expected = 'tac';
```
Comment (matthewmorgan) [inline exercises/reverse-string/reverse-string.spec.js]: For this test, the canonical data tests reversing `robot` instead of `cat`.

Please change this test to match the canonical one.

File: exercises/reverse-string/reverse-string.spec.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+var reverseString = require('./reverse-string');
+
+describe('ReverseString', function () {
+  it('empty string', function () {
+    var expected = '';
+    var actual = reverseString('');
+    expect(actual).toEqual(expected);
+  });
+
+  it('a word', function () {
```
Comment (matthewmorgan) [inline exercises/reverse-string/reverse-string.spec.js]: All `it()` blocks except the first one should be changed to `xit()`.  This allows the student to pass one test at a time.

**Comment by eliaahadi** [comment]
Ok made the change you requested I think, please review again and thanks for all your help!

**Comment by matthewmorgan** [comment]
Nice!  Congratulations on a job well done.

**Comment by eliaahadi** [comment]
Great thanks! Will this show up on the exesrcism website or are there more steps I need to do?

**Comment by matthewmorgan** [comment]
@eliaahadi I believe each track is deployed once per day by an automated process.  

Normally changes show up within 24 hours, but let me know if you don't see the change by this time tomorrow and I can check in to it.

**Comment by matthewmorgan** [comment]
@eliaahadi I have not seen this deployed yet.  I put in [an issue](https://github.com/exercism/exercism.io/issues/3703) to find out why.

**Comment by matthewmorgan** [comment]
OK!  Katrina deployed the site, and your contribution is live.

Congratulations again. 🎉

**Comment by eliaahadi** [comment]
Great thanks again Matt! It took awhile, but we got it done. Happy new year!
